test/bufconn/bufconn.go

```diff
@@ -232,11 +276,26 @@ func (c *conn) Close() error {
 	return err2
 }
 
-func (*conn) LocalAddr() net.Addr                  { return addr{} }
-func (*conn) RemoteAddr() net.Addr                 { return addr{} }
-func (c *conn) SetDeadline(t time.Time) error      { return fmt.Errorf("unsupported") }
-func (c *conn) SetReadDeadline(t time.Time) error  { return fmt.Errorf("unsupported") }
-func (c *conn) SetWriteDeadline(t time.Time) error { return fmt.Errorf("unsupported") }
+func (c *conn) SetDeadline(t time.Time) error {
+	if err := c.SetReadDeadline(t); err != nil {
+		return err
+	}
+
+	return c.SetWriteDeadline(t)
+}
+
+func (c *conn) SetReadDeadline(t time.Time) error {
+	c.Reader.(*pipe).rtimeout = t
```

Comment:
Per the `net.Conn` documentation, this is supposed to affect all pending reads (and writes) as well as future ones.  So I think something else needs to happen here with `c.Reader.rtimer`.